tephra-hbase-compat-1.1-base/src/main/java/org/apache/tephra/hbase/txprune/PruneUpperBoundWriter.java

```diff
@@ -76,18 +76,21 @@ private void startFlushThread() {
     flushThread = new Thread("tephra-prune-upper-bound-writer") {
       @Override
       public void run() {
-        while (!isInterrupted()) {
+        while (!isInterrupted() && !stopped) {
           long now = System.currentTimeMillis();
           if (now > (lastChecked + pruneFlushInterval)) {
-            if (shouldFlush.compareAndSet(true, false)) {
+            if (!pruneEntries.isEmpty()) {
               // should flush data
               try {
-                dataJanitorState.savePruneUpperBoundForRegion(regionName, pruneUpperBound.get());
+                while (pruneEntries.firstEntry() != null) {
+                  Map.Entry<ByteBuffer, Long> firstEntry = pruneEntries.firstEntry();
+                  dataJanitorState.savePruneUpperBoundForRegion(firstEntry.getKey().array(), firstEntry.getValue());
+                  // We can now remove the entry only if the key and value match with what we wrote since it is
+                  // possible that a new pruneUpperBound for the same key has been added
+                  pruneEntries.remove(firstEntry.getKey(), firstEntry.getValue());
+                }
               } catch (IOException ex) {
-                LOG.warn("Cannot record prune upper bound for region " + regionNameAsString + " in the table " +
-                           pruneStateTable.getNameWithNamespaceInclAsString() + " after compacting region.", ex);
-                // Retry again
-                shouldFlush.set(true);
+                LOG.warn("Cannot record prune upper bound for a region.", ex);
```

Comment:
It would be good to add prune state table name to the log message, so that it can help in figuring out mis-configurations if any